### DISCUSSION TITLE
Variables/references in documentation

### DISCUSSION THREAD

**Discussion by malakeel** [testingisdocumenting/znai#1034] [General]
I am wondering about the way to use references in the documentation. For example, if I have a word that I need to use in multiple places where copy/paste is not the best way. An example could be an entity name that could change, a software version number .. etc. 

What is the best way to do this other than using search and replace ?

**Comment by MykolaGolubyev** [comment]
Good question. Something I had in a backlog for a while. 
there is a plug-in include-markdown that includes a different markdown file in the current one. If we add an inline version of it then it may work for your use case.

**Reply by malakeel** [reply]
Yes, the "inline" would do the trick. I overlooked that part

**Comment by malakeel** [comment]
@MykolaGolubyev 

I appreciate the efforts, and the answer. I have read somewhere that znai uses [FreeMarker]. If this is correct then something like the following should work:

`
<#assign x = 1> 
${x}

`
However, this was not the case for me, or at least I missed something. I guess it depends on how FreeMarker was used in the code.
Following up with your suggestion, here's what I did and it worked (others in the same position can use this method):

* create a directory somewhere. I created it in the root of the doc `variables`.
* add `./variables` to lookup-paths file.
* create a file like  `variables/maven-version.md` with the content I want to use. Let's say `3.8.4`.
* In my mark-down I used:
:include-markdown: maven-version.md 

And the variable is displayed properly.

This gets the job done. May be not obvious, nor convenient, but it is okay. Thank you for your help

[FreeMarker]:  https://testingisdocumenting.org/znai/layout/templates#extract-a-pattern

**Reply by malakeel** [reply]
I guess, the replacement has to be on its own line. Otherwise it wont work if in the middle of text. 
Hum !

**Reply by MykolaGolubyev** [reply]
https://github.com/testingisdocumenting/znai/pull/1036

**Comment by MykolaGolubyev** [comment]
I agree that include-markdown may not be obvious. Will work on adding inline support. 

Could you clarify what you want to achieve with template? 

Thanks for starting this discussion. If you have suggestions for the plugins - do let me know

**Reply by malakeel** [reply]
@MykolaGolubyev 
I need to only replace a variable in the markdown. I can't imaging this uncommon requirement by other authors. So for example, you may want to replace an entity name across all content, and you may use a variable for this. 

In the znai documentation, there's a [page about references], and this page is not very clear on how to use references. In the example I gave earlier, where I want to replace a version of the software (or may be the name of the software) across all content, like in maven version case.

[page about references]: https://testingisdocumenting.org/znai/example-references/domain

**Reply by MykolaGolubyev** [reply]
The page about reference is an example of how to link code snippets identifiers to Reference docs. I will need to put an extra info to avoid confusion.

I hear you about var replace. Seems like a good idea to add a native support. Something like 
```
some text `:var: MY_VAR` another text
```
but then could be tough to use it inside things like `badge` plugin.

The way our docs and docs of some of the folks deal with things like version is by using regular file insert. E.g. maven resource plugin substitutes placeholders with variables (maven version) and then znai refers preprocessed files.

Include-template plugin that uses Free Mark template language is also an option.

What is your specific use case? May help me provide you a work around while I am thinking about in-place variables support.

**Reply by malakeel** [reply]
This is exactly what I am looking for. I am not sure I get you about badge plugin conflict. If replacing variables is done during "preprocessing" stage, then the resulting text should be fine to go through additional plugin. For my needs, I don't think I will need to use the badge, but others may need it. 

So here's a possible usage (just as pseudo code), assuming MY_VAR has the value ZNAI.

```
some text  `:badge:  :var: MY_VAR ` another text 
```
After preprocessing step:
```
some text  `:badge:  ZNAI ` another text 
```

Which will finally be processed by badge plugin, producing

some text  **ZNAI**  another text

Now, to avoid ambiguity, and to maintain the good practice of separating processing logic from presentation and styling, we may want to use different notation. Here FreeMarker can help us, so if we do:

'''
some text  `:badge:  ${MY_VAR} ` another text 
'''

and let FreeMarker take control early, then we avoid the conflict with other plugins. I am not sure at what stage currently FreeMarker is invoked (if at all), but I prefer to distinguish between using FreeMarker for things related to expression evaluation/replacement ... etc, and using FreeMarker for doing presentation related tasks. 
Not sure if this separation already exists in the plugins. For example I don't know if you already have per-processing plugins, and presentation plugins for markdown related tasks. After all, if what you meant by native support is allow all FreeMarker stuff to be processed early, then I think this is good idea.

**Reply by MykolaGolubyev** [reply]
FreeMarker support is only for `template` plugin at the moment: https://testingisdocumenting.org/znai/layout/templates

pre-processing the whole file with FreeMarker most likely will change the default behavior of markdown. E.g. ${myvar} snippet will be replaced when a user wants to render var name itself and will require escaping. I am reluctant to change the default behavior even though it will make things so much simpler.

The reason I proposed ``` `:var:` ``` is that it will leverage Znai plugins system. There are three plugin types: include, fenced block and inlined code. ``` `:<plugin-name>:` ``` is the type of inlined code plugins. They are processed only once. 

How do you plan to set a variable value? I am thinking that `template` plugin may do the job for you. Or we can make it so template plugin has an inlined version and also automatically resolved env-vars and system props without the need to pass them explicitly.

**Reply by MykolaGolubyev** [reply]
@malakeel let me know if inline template plugin is something you will find useful

**Reply by malakeel** [reply]
@MykolaGolubyev 
Thank you for getting back. I think it is useful, but I prefer if you don't take my opinion only. I can get around this temporarily by running some other preprocessing step. It is not urgent. I prefer if you think about the best way to do it, and how to address most case cases. You know better than anyone about the program, so you can handle it in the most generic way.
From what you described, I think it makes sense to use inline template plugin. But at the same time, preprocessing with Freemarker makes sense. The program is in early stages, and unlikely to break something if you change the default behavior. If changing the default behavior improves things for the long run, then I would do it there (if I were you).

**Reply by MykolaGolubyev** [reply]
Appreciate the thought process. 
so how do you plan to pass the variable values?

If you have time, could you please fill a bit of info here on how do you use or plan to use znai: https://github.com/testingisdocumenting/znai/discussions/824

**Reply by MykolaGolubyev** [reply]
@malakeel were you able to figure things out? Sorry I sort of dropped the ball on this one

**Reply by MykolaGolubyev** [reply]
@malakeel circling back here, do you still need help on this?
